content/blog/2022/2022-08-03-feature-flags-in-nodejs-backend-server-with-launchdarkly.md

```diff
@@ -0,0 +1,720 @@
+
+Now we can take a brief look at the JSON type flags as these feature flag values are pretty open-ended. LaunchDarkly supports a JSON type right out of the box. This allows us to pass Object and Array data structures to our application which can then be used to implement lightweight administrative and operational functionality in our web application.
+
+Let’s consider that we have a server that watches the network IPs that are trying to access the server and somehow it would like to check for blacklisted IPs and then block them from our server. So if we think about it on a very bare minimum we would need the following:
+
+- a database to store the IP address.
+- an Admin UI to update those IP addresses periodically.
+- a User Access Control around those User Interfaces.
+- a data-access layer as an abstraction to update the database.
+- finally, a REST API to integrate with the UI.
+
+All of this looks tedious and need a lot of effort to maintain just a list of IP address. Well, LaunchDarkly provides us with all of this functionality with almost no effort at all. We can just define a flexible data structure and pass those values to be fetched dynamically. In this case, we can create an Array of IP addresses as part of JSON variation. This helps us to maintain a history of IP-address configuration rather than changing the same value. A type of this feature flag would look something like:
+
+{{% image alt="JSON multivariate feature flag" src="images/posts/nodejs-backend-launchdarkly/IP_Blacklist.png" %}}
+
+Next we will define the code to consume this feature flag variation dynamically and block the IP:
+
+```javascript
+import chalk from 'chalk';
+import LaunchDarkly from 'launchdarkly-node-server-sdk';
+
+const LD_SDK_KEY = 'sdk-d2432dc7-e56a-458b-9f93-0361af47d578';
+const launchDarklyClient = LaunchDarkly.init( LD_SDK_KEY );
+ 
+// We're just going to mock some traffic to a
+// mock function that handles mock request /
+// response structures mock all the things!
+(async function sendMockHttpRequest() { 
+	try {
+		await launchDarklyClient.waitForInitialization(); 
+		await mockHandleHttpRequest({
+			ip: "192.168.100.40"
+		}); 
+	} catch ( error ) { 
+		console.warn( "Error sending mock HTTP request." );
+		console.error( error ); 
+	} 
+	setTimeout( sendMockHttpRequest, 1000 );
+ 
+})();
+
+async function mockHandleHttpRequest( request ) {
+ 
+	// Get the collection of blocked IP-addresses. Internally, we've defined our feature
+	// flag value as a JSON (JavaScript Object Notation) type. This means that
+	// LaunchDarkly automatically handles the stringification, streaming, and parsing of
+	// it for us. In other words, this .variation() call doesn't return a JSON payload -
+	// it returns the original data structure that we provided in the LD dashboard.
+	var ipBlacklist = await launchDarklyClient.variation(
+		'ip-blacklisted-list',
+		{
+			key: 'mob-b9d6d4d4-4300-46fa-9b13-d9eac89f9794' // The static "user" for this task.
```

Comment:
As said above ... I don't think we need a user at all. Instead, we can pass a "global" user like `global` or something.